### DISCUSSION TITLE
Reorganize git hooks

### DISCUSSION THREAD

**Discussion by tcmetzger** [bokeh/bokeh#11431] [Build / CI]
Currently, we encourage people to create and install git hooks manually when setting up a dev environment: https://docs.bokeh.org/en/latest/docs/dev_guide/setup.html#configuring-git

These steps only work on Linux/macOS (not windows), require a lot of manual setup (creating, copying and `chmod`ing files), and take up a lot of space in the dev guide setup instructions. Those git hooks are great to have, so it would be helpful to have an easier way to distribute and set up those files.

There are several solutions available for us to consider:

- I have seen several projects use https://pre-commit.com/ (and used it myself as well). This package is available on conda-forge and is based on a YAML config file to include in the repo. In most cases, you just need to run `pre-commit install` to get things set up locally. It is based on a set of pre-configured hooks (https://pre-commit.com/hooks.html) but can also be customized further.
- An interesting npm-based approach seems to be Husky: https://typicode.github.io/husky/#/
- Another option is to provide our own install scripts that would copy pre-configured hooks into the `.git/hooks` folder.

That's just some ideas, I'm sure there are more ways to make Git hooks work more easily for contributors!

**Comment by kir0ul** [comment]
100% agree that it would be nice to remove those manual steps and replace them by one simple command like `pre-commit install` or something else!
If I'm not mistaken, I believe one advantage of `pre-commit` over other solutions is that it supports [other languages](https://pre-commit.com/#supported-languages) than only JS.

**Comment by bryevdv** [comment]
I just went through the Dask dev setup an they use `pre-commit` and it was super simple:

https://docs.dask.org/en/latest/develop.html#code-formatting

 It looks like they are using some "third party" hooks that get referenced in the configurtation:

https://github.com/dask/distributed/blob/5dc591bbdd4427fe49fe90338a34fc85ee35f2c9/.pre-commit-config.yaml

But it looks like we could define custom hooks as well, so we could run e.g. our full codebase tests or our branch protection check as `pre-push` hooks. 

TLDR I would be +1 on using `pre-commit` stating with two existing manual `pre-push` hooks in the docs, then discussing any extra hooks to add as pre-commit e.g. running `isort` or a quick linter like `pyflakes`

**Comment by tcmetzger** [comment]
@kir0ul Would you be interested in starting a PR to add pre-commit to the repo? Like @bryevdv said, as a first step it would be great to transfer the git hooks described in https://docs.bokeh.org/en/latest/docs/dev_guide/setup.html#git-hooks to pre-commit. I'd be happy to help with that, of course!

**Reply by kir0ul** [reply]
Yes, I can give it a try! :slightly_smiling_face:

**Reply by tcmetzger** [reply]
Let me know how it goes - and how I can help! 😀

**Comment by kir0ul** [comment]
Started here: https://github.com/bokeh/bokeh/pull/11442
A few comments:
- The `pre-push` should probably have to be rewritten as a Python script so that it can be cross platforms.
- AFAIK there's no cross platform way of getting the default editor and opening all the files with conflicts, so I'm not sure if there's a way to remove the manual step of setting the [Git Aliases](https://docs.bokeh.org/en/test/docs/dev_guide/setup.html#git-aliases), but maybe someone else knows better?
